### PR TITLE
__hgt2 fix, env vars to individually disable fusion ops, MNIST testcase fix

### PR DESCRIPTION
**Pull request by deven-amd**
PR Contents:
-------------------------------
workaround for the issue where __hgt2 returns -1 (instead of 1) for true

-------------------------------
 ROCm Fusion Pass updates

1. change the pass to traverse the TF graph individually for each fusion

2. add env vars to disable fusions on an individual basis (to help with debug / perf measurements)

   By default, ROCm FUSION is turned OFF
     - Setting the env var TF_ROCM_FUSION_ENABLE (to any value) turns ALL the fusion ops ON

   When the env var TF_ROCM_FUSION_ENABLE is set
   - set `TF_ROCM_FUSION_DISABLE_<type>` to disable `<type>` fusion, where <type> can be
     - `CBA` : Conv+Bias+Actv
     - `BNA` : BatchNorm+Actv
     - `ADDRELU` : Add+Relu
     - `ADDNRELUGRAD` - AddN+ReluGrad

-------------------------------

 AddRelu Kernel implementation updates

1. added to check to error out when assumptions regarding the shape of the input tensors are not correct

2. added a simple "broadcast" version of the AddRelu Kernel to make the models/image/tutorial/mnist testcase work correctly
   the mnist testcase was giving incorrect results because the "Add" operation requires broadcasting.
   the "Add" operation in the resenet50 model is element-wise (i.e. does not require any broadcast) so this change does not affect it all.
-------------------------------

### CODE REVIEW

**Comment by sunway513** [comment]
@deven-amd , do I need to pick this PR to r1.11-rocm-fusion branch?

**Comment by deven-amd** [comment]
> @deven-amd , do I need to pick this PR to r1.11-rocm-fusion branch?

You do not need to. No functional / performance changes that will affect resnet50.

On the flip side, one of the commits will allow you to disable individual fusions. It might help you with doing performance measurements with selective fusions turned on/off, in case you need to do something like that.

File: tensorflow/core/graph/gpu_fusion_pass.cc [hunk 1]
```diff
@@ -421,30 +412,27 @@ bool ROCmFusionPass::RunPass(Graph* graph) {
...
+
+  for (auto& fusion : fusions) {
+    std::vector<Node*> order;
+    GetPostOrder(*graph,
+                 &order);  // This will give us reverse topological sort.
+
+    std::set<const Node*> fused_nodes;
+
+    for (const Node* n : order) {
+      if (fused_nodes.count(n)) {
```
Comment (whchung) [inline tensorflow/core/graph/gpu_fusion_pass.cc]: @deven-amd in this new version, wouldn't `fused_nodes` be always empty and this check always be `false`?

Comment (whchung) [reply tensorflow/core/graph/gpu_fusion_pass.cc]: let me take back my above question.

File: tensorflow/core/graph/gpu_fusion_pass.cc [hunk 2]
```diff
@@ -456,27 +444,39 @@ bool ROCmFusionPass::RunPass(Graph* graph) {
...
-    std::set<const Node*>& fused_nodes) {
-  fusions.emplace_back(
-      new ROCmFusionOpConvolutionBiasBatchNormActivation(g, fused_nodes));
-  fusions.emplace_back(
-      new ROCmFusionOpConvolutionBiasActivation(g, fused_nodes));
-  fusions.emplace_back(new ROCmFusionOpBatchNormActivation(g, fused_nodes));
-  fusions.emplace_back(new ROCmFusionOpAddRelu(g, fused_nodes));
-  fusions.emplace_back(new ROCmFusionOpAddNReluGrad(g, fused_nodes));
+    std::vector<std::unique_ptr<ROCmFusionOpBase> >& fusions, Graph* g) {
+  if (!getenv("TF_ROCM_FUSION_DISABLE_CBNA")) {
```
Comment (whchung) [inline tensorflow/core/graph/gpu_fusion_pass.cc]: It seems in TF there are utility functions `ReadXXXFromEnvVar`. Could you help use that? Perhaps `ReadBoolFromEnvVar` would be a good choice.

Comment (deven-amd) [reply tensorflow/core/graph/gpu_fusion_pass.cc]: make sense...will make that change in the next commit

Comment (whchung) [reply tensorflow/core/graph/gpu_fusion_pass.cc]: the test is a success. once you make the change I'll take it.
